src/osd/PrimaryLogPG.cc

```diff
@@ -3717,6 +3717,39 @@ int PrimaryLogPG::do_xattr_cmp_str(int op, string& v1s, bufferlist& xattr)
   }
 }
 
+int PrimaryLogPG::do_extent_cmp(OpContext *ctx, OSDOp& osd_op)
+{
+  ceph_osd_op& op = osd_op.op;
+  vector<OSDOp> read_ops(1);
+  OSDOp& read_op = read_ops[0];
+  int result = 0;
+
+  read_op.op.op = CEPH_OSD_OP_SYNC_READ;
+  read_op.op.extent.offset = op.extent.offset;
+  read_op.op.extent.length = op.extent.length;
+  read_op.op.extent.truncate_seq = op.extent.truncate_seq;
+  read_op.op.extent.truncate_size = op.extent.truncate_size;
+
+  result = do_osd_ops(ctx, read_ops);
+  if (result < 0) {
+    derr << "do_extent_cmp do_osd_ops failed " << result << dendl;
+    return result;
+  }
+
+  if (read_op.outdata.length() != osd_op.indata.length())
+    return -EINVAL;
+
+  for (uint64_t p = 0; p < osd_op.indata.length(); p++) {
+    if (read_op.outdata[p] != osd_op.indata[p]) {
+      ::encode(p, osd_op.outdata);
+      osd_op.outdata.claim_append(read_op.outdata);
```

Comment:
Thanks for the feedback. This seems like a nice simplification to the API/proto - I'll work on changes to implement this.

It'll cause a few headaches for those who hastily picked up the old patch-set without upstream blessing (yours truly), but it should be easy enough to overcome by carrying an out-of-tree -EILSEQ handler on the client side, in addition to the miscompare-on-positive-return logic.